perma_web/warc_server/pywb_config.py

```diff
@@ -17,58 +15,149 @@
 from django.core.exceptions import DisallowedHost
 from django.core.cache import cache as django_cache
 
+from pywb.cdx.cdxserver import CDXServer
 from pywb.cdx.cdxsource import CDXSource
 from pywb.framework import archivalrouter
 from pywb.framework.wbrequestresponse import WbResponse
+from pywb.framework.memento import MementoResponse, LINK_FORMAT
 from pywb.rewrite.wburl import WbUrl
 from pywb.utils.loaders import BlockLoader, LimitReader
-from pywb.warc.cdxindexer import write_cdx_index
 from pywb.webapp.handlers import WBHandler
 from pywb.webapp.query_handler import QueryHandler
 from pywb.webapp.pywb_init import create_wb_handler
 from pywb.webapp.views import add_env_globals
+from pywb.webapp.pywb_init import create_wb_router
+from pywb.utils.wbexception import NotFoundException
 
-from perma.models import Link
+from perma.models import CDXLine, Asset, Link
+
+# Assumes post November 2013 GUID format
+GUID_REGEX = r'([a-zA-Z0-9]+(-[a-zA-Z0-9]+)+)'
 
 
 # include guid in CDX requests
-class Route(archivalrouter.Route):
+class PermaRoute(archivalrouter.Route):
     def apply_filters(self, wbrequest, matcher):
-        wbrequest.custom_params['guid'] = matcher.group(1)
+        """Parse the GUID and find the CDXLine in the DB"""
+
+        guid = matcher.group(1)
+        urlkey = surt(wbrequest.wb_url_str)
+
+        try:
+            # This will filter out links that have user_deleted=True
+            link = Link.objects.get(guid=guid)
+        except Link.DoesNotExist:
+            raise NotFoundException()
+
+        try:
+            line = CDXLine.objects.get(urlkey=urlkey,
+                                       asset__link_id=guid)
+        # Legacy archives didn't generate CDXLines during
+        # capture so generate them on demand if not found
+        except CDXLine.DoesNotExist:
```

Comment:
Does this distinguish between an index we haven't created yet, and an asset we really don't have? It's common for a complex playback like /ABCD-1234/http://cnn.com/ to trigger a bunch of requests like /ABCD-1234/http://assets.cnn.com/something-we-missed.css -- sometimes an optimistic javascript will even kick off continual requests for the same missing asset. We don't want to be re-running the indexing step each time.

Unless I'm missing something, maybe we need to add an asset.cdx_created field to avoid extra indexing? Or we could run the index only if there are no CDXLines at all -- not sure if that opens up weird edge cases though?

(Let's chat when you have a minute about the best way to normalize archives and assets in general -- I don't love what we have now, but not sure if there's a better way.)